### PR TITLE
[CALCITE-3587] RexBuilder may lose decimal fraction for creating literal with DECIMAL type

### PR DESCRIPTION
**Pull request by yanlin-Lynn**
DECIMAL is fixed point type, when creating RexLiteral with DECIMAL type,
should not only use the integer part, need the decimal fraction part as well

### CODE REVIEW

**Comment by amaliujia** [comment]
Good catch. So the Decimal hits `new BigDecimal(((Number) o).longValue())` which causes the fraction loss.

**Comment by julianhyde** [comment]
See comments in JIRA issue.

**Comment by yanlin-Lynn** [comment]
Update PR according to comments.

File: core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java [hunk 1]
```diff
@@ -414,6 +414,23 @@ private void checkDate(RexNode node) {
     assertThat(literal.getValue3(), nullValue());
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3587">[CALCITE-3587]
+   * RexBuilder may lose decimal fraction for creating literal with DECIMAL type</a>.
+   */
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java]: lose decimal fraction -> loose decimal precision ?

File: core/src/main/java/org/apache/calcite/rex/RexBuilder.java [hunk 2]
```diff
@@ -1517,6 +1517,8 @@ private static Object clean(Object o, RelDataType type) {
       if (o instanceof BigDecimal) {
         return o;
       }
+      assert !(o instanceof Float || o instanceof Double)
+          : o.getClass().getCanonicalName() + " is not compatible with " + type.getSqlTypeName();
       return new BigDecimal(((Number) o).longValue());
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rex/RexBuilder.java]: What do you mean by "s not compatible with" ?

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/rex/RexBuilder.java]: Will it better to include a reminder that use `makeExactLiteral` to create Decimal literal by a float or double input?

File: core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java (hunk 1 above)
Comment (yanlin-Lynn) [reply core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java]: no. It's lose, not loose.
Loose is an adjective that means “not tight.”

Comment (danny0405) [reply core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java]: What do you mean by "fraction" ?

Comment (yanlin-Lynn) [reply core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java]: "decimal fraction", means the non integer part . for "12.95", means the ".95" part.

**Pull request merged**